yarn-project/simulator/src/avm/journal/public_storage.ts

```diff
@@ -0,0 +1,149 @@
+import { Fr } from '@aztec/foundation/fields';
+
+import type { PublicStateDB } from '../../index.js';
+
+/**
+ * A class to manage public storage reads and writes during a contract call's AVM simulation.
+ * Maintains a storage write cache, and ensures that reads fall back to the correct source.
+ * When a contract call completes, its storage cache can be merged into its parent's.
+ */
+export class PublicStorage {
+  /** Cached storage writes. */
+  private cache: PublicStorageCache;
+  /** Parent's storage cache. Checked on cache-miss. */
+  private readonly parentCache: PublicStorageCache | undefined;
+  /** Reference to node storage. Checked on parent cache-miss. */
+  private readonly hostPublicStorage: PublicStateDB;
+
+  constructor(hostPublicStorage: PublicStateDB, parent?: PublicStorage) {
+    this.hostPublicStorage = hostPublicStorage;
+    this.parentCache = parent?.cache;
+    this.cache = new PublicStorageCache();
+  }
+
+  /**
+   * Get the pending storage.
+   */
+  public getCache() {
+    return this.cache;
+  }
+
+  /**
+   * Read a value from storage.
+   * 1. Check cache.
+   * 2. Check parent's cache.
+   * 3. Fall back to the host state.
+   * 4. Not found! Value has never been written to before. Flag it as non-existent and return value zero.
+   *
+   * @param storageAddress - the address of the contract whose storage is being read from
+   * @param slot - the slot in the contract's storage being read from
+   * @returns exists: whether the slot has EVER been written to before, value: the latest value written to slot, or 0 if never written to before
+   */
+  public async read(storageAddress: Fr, slot: Fr): Promise<[/*exists=*/ boolean, /*value=*/ Fr]> {
+    // First try check this storage cache
+    let value = this.cache.read(storageAddress, slot);
+    // Then try parent's storage cache (if it exists / written to earlier in this TX)
+    if (!value && this.parentCache) {
+      value = this.parentCache?.read(storageAddress, slot);
+    }
+    // Finally try the host's Aztec state (a trip to the database)
+    if (!value) {
+      value = await this.hostPublicStorage.storageRead(storageAddress, slot);
+    }
+    // if value is undefined, that means this slot has never been written to!
+    const exists = value !== undefined;
+    const valueOrZero = exists ? value : Fr.ZERO;
+    return Promise.resolve([exists, valueOrZero]);
```

Comment:
Not at the moment, but it'll be relevant to the public kernel as it's going to need to do a non-membership check instead of check membership of a specific value. Ultimately it'll need to be in the trace output.